test-framework/emulator_backend.go

```diff
@@ -0,0 +1,498 @@
+			Error: err,
+		}
+	}
+
+	if result.Error != nil {
+		return &stdlib.TransactionResult{
+			Error: result.Error,
+		}
+	}
+
+	return &stdlib.TransactionResult{}
+}
+
+func (e *EmulatorBackend) CommitBlock() error {
+	// Reset the transaction offset for the current block.
+	e.blockOffset = 0
+
+	_, err := e.blockchain.CommitBlock()
+	return err
+}
+
+func (e *EmulatorBackend) DeployContract(
+	name string,
+	code string,
+	account *stdlib.Account,
+	args []interpreter.Value,
+) error {
+
+	const deployContractTransactionTemplate = `
+	    transaction(%s) {
+		    prepare(signer: AuthAccount) {
+			    signer.contracts.add(name: "%s", code: "%s".decodeHex()%s)
+		    }
+	    }`
+
+	code = e.replaceImports(code)
+
+	hexEncodedCode := hex.EncodeToString([]byte(code))
+
+	inter, err := newInterpreter()
+	if err != nil {
+		return err
+	}
+
+	cadenceArgs := make([]cadence.Value, 0, len(args))
+
+	var txArgsBuilder, addArgsBuilder strings.Builder
+
+	for i, arg := range args {
+		cadenceArg, err := runtime.ExportValue(arg, inter, interpreter.ReturnEmptyLocationRange)
+		if err != nil {
+			return err
+		}
+
+		if i > 0 {
+			txArgsBuilder.WriteString(", ")
+		}
+
+		txArgsBuilder.WriteString(fmt.Sprintf("arg%d: %s", i, cadenceArg.Type().ID()))
+		addArgsBuilder.WriteString(fmt.Sprintf(", arg%d", i))
+
+		cadenceArgs = append(cadenceArgs, cadenceArg)
+	}
+
+	script := fmt.Sprintf(
+		deployContractTransactionTemplate,
+		txArgsBuilder.String(),
+		name,
+		hexEncodedCode,
+		addArgsBuilder.String(),
+	)
+
+	tx := e.newTransaction(script, []common.Address{account.Address})
+
+	for _, arg := range cadenceArgs {
+		err = tx.AddArgument(arg)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = e.signTransaction(tx, []*stdlib.Account{account})
+	if err != nil {
+		return err
+	}
+
+	err = e.blockchain.AddTransaction(*tx)
+	if err != nil {
+		return err
+	}
+
+	// Increment the transaction sequence number offset for the current block.
+	e.blockOffset++
+
+	result := e.ExecuteNextTransaction()
+	if result.Error != nil {
+		return result.Error
+	}
+
+	return e.CommitBlock()
+}
+
+func (e *EmulatorBackend) ReadFile(path string) (string, error) {
+	if e.importResolver == nil {
+		return "", ImportResolverNotProvidedError{}
+	}
+
+	return e.importResolver(common.StringLocation(path))
+}
+
+// newBlockchain returns an emulator blockchain for testing.
+func newBlockchain(opts ...emulator.Option) *emulator.Blockchain {
+	b, err := emulator.NewBlockchain(
+		append(
+			[]emulator.Option{
+				emulator.WithStorageLimitEnabled(false),
+			},
+			opts...,
+		)...,
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	return b
+}
+
+func (e *EmulatorBackend) UseConfiguration(configuration *stdlib.Configuration) {
+	e.configuration = configuration
+}
+
+// newInterpreter creates an interpreter instance needed for the value conversion.
+//
+func newInterpreter() (*interpreter.Interpreter, error) {
+	// TODO: maybe re-use interpreter? Only needed for value conversion
```

Comment:
Re-using should be fine if the only thing that's done is using it for non-interpretation, like value conversion. Then again, it shouldn't be expensive either.

Reply:
yeah, I think it's fine to keep it for now. We can improve it later, if needed